buildSrc/src/main/kotlin/fat-jar.gradle.kts

```diff
@@ -0,0 +1,99 @@
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import io.spine.internal.gradle.publish.SpinePublishing
+
+plugins {
+    java
+    `maven-publish`
+    com.github.johnrengelman.shadow
+}
+
+/**
+ * Creates a custom publication called `fatJar`, taking the output of `tasks.shadowJar`
+ * as the sole publication artifact.
+ *
+ * This kind of publishing works in combination with exclusion of standard publishing for this
+ * module specified in the root project under [SpinePublishing.modulesWithCustomPublishing].
+ */
+publishing {
+    publications {
+        // The publishing settings from the root project.
+        val spinePublishing = rootProject.the<SpinePublishing>()
+        create("fatJar", MavenPublication::class) {
+            groupId = project.group.toString()
+            artifactId = spinePublishing.artifactId(project)
+            version = project.version.toString()
+            artifact(tasks.shadowJar)
+        }
+    }
+}
+
+tasks.publish {
+    dependsOn(tasks.shadowJar)
+}
+
+tasks.shadowJar {
+    exclude(
+        /**
+         * Excluding this type to avoid it being located in the fat JAR.
+         *
+         * Locating this type in its own `io:spine:protodata` artifact is crucial
+         * for obtaining proper version values from the manifest file.
+         * This file is only present in `io:spine:protodata` artifact.
+         */
+        "io/spine/protodata/gradle/plugin/Plugin.class",
+        "META-INF/gradle-plugins/io.spine.protodata.properties",
+
+        /**
+         * Exclude Gradle types to reduce the size of the resulting JAR.
+         *
+         * Those required for the plugins are available at runtime anyway.
+         */
+        "org/gradle/**",
+
+        /**
+         * Remove all third-party plugin declarations as well.
+         *
+         * They should be loaded from their respective dependencies.
+         */
+        "META-INF/gradle-plugins/com**",
+        "META-INF/gradle-plugins/net**",
+        "META-INF/gradle-plugins/org**")
+
+    isZip64 = true  /* The archive has way too many items. So using the Zip64 mode. */
+    archiveClassifier.set("")    /** To prevent Gradle setting something like `osx-x86_64`. */
```

Comment:
If this is applied to all of our fat JARs, some of them (namely, artifacts defined in ProtoData and McJava) will probably lose their classifiers used currently.

Another thing is that we reference some of artifacts programmatically (McJava does that, at least, from what I recall). Therefore, any changes to coordinates, such as setting/removing the classifiers, should also be reflected in our code. 

That's something to consider.

Reply:
@armiol, I suppose if those modules define their own `tasks.shadowJar`, they can "override" this behaviour when setting the classifier explicitly. They should be evaluated _after_ those tasks that are added by a script plugin. Don't they?

Reply:
alexander-yevsyukov probably so, but with all these `afterSmth` tricks I would not be so sure. 

In any case, I would mention the classifier configuration in docs.